File: src/python/xerxes/tui/context_bar.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Context-utilization bar.

Render a filled/empty block meter that visualises ``used / window``
token utilization for the footer."""

from __future__ import annotations

FILLED_BLOCK = "█"
HALF_BLOCK = "▌"
EMPTY_BLOCK = "░"


def context_bar(
    *,
    used: int,
    window: int,
    width: int = 24,
    filled: str = FILLED_BLOCK,
    half: str = HALF_BLOCK,
    empty: str = EMPTY_BLOCK,
) -> str:
    """Return a ``width``-character meter.

    Halves a block when utilization falls between integer fill counts
    so very low usage still produces a visible mark."""
    if width <= 0:
        return ""
    window = max(1, int(window))
    used = max(0, int(used))
    cells = (used / window) * width if used < window else width
    full = int(cells)
    has_half = (cells - full) >= 0.5
    if full > width:
        full = width
        has_half = False
    extra = 1 if has_half else 0
    return (filled * full) + (half if has_half else "") + (empty * (width - full - extra))


def context_bar_with_pct(
    *,
    used: int,
    window: int,
    width: int = 24,
    show_pct: bool = True,
) -> str:
    """Return :func:`context_bar` plus a right-aligned ``NN.N%`` suffix.

    Set ``show_pct=False`` to suppress the percentage and get the raw bar."""
    bar = context_bar(used=used, window=window, width=width)
    if not show_pct:
        return bar
    pct = (used / window * 100) if window > 0 else 0
    return f"{bar} {pct:5.1f}%"


__all__ = ["EMPTY_BLOCK", "FILLED_BLOCK", "HALF_BLOCK", "context_bar", "context_bar_with_pct"]

File: src/typescript/src/ui/lib/contextBar.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.
/** Compact context-window utilization meters for status surfaces. */

export const FILLED_BLOCK = '█'
export const HALF_BLOCK = '▌'
export const EMPTY_BLOCK = '░'

export interface ContextBarOptions {
  empty?: string
  filled?: string
  half?: string
  used: number
  width?: number
  window: number
}

export interface ContextBarWithPctOptions extends ContextBarOptions {
  showPct?: boolean
}

const integer = (value: number, label: string): number => {
  if (!Number.isFinite(value)) {
    throw new RangeError(`${label} must be finite`)
  }

  return Math.trunc(value)
}

/**
 * Renders a fixed-cell meter for used context tokens.
 *
 * A half cell makes low utilization visible while clamping any overfull
 * context to the provided width.
 */
export const contextBar = ({
  empty = EMPTY_BLOCK,
  filled = FILLED_BLOCK,
  half = HALF_BLOCK,
  used,
  width = 24,
  window
}: ContextBarOptions): string => {
  const safeWidth = integer(width, 'width')

  if (safeWidth <= 0) {
    return ''
  }

  const safeWindow = Math.max(1, integer(window, 'window'))
  const safeUsed = Math.max(0, integer(used, 'used'))
  const cells = safeUsed < safeWindow ? (safeUsed / safeWindow) * safeWidth : safeWidth
  let full = Math.trunc(cells)
  let hasHalf = cells - full >= 0.5

  if (full > safeWidth) {
    full = safeWidth
    hasHalf = false
  }

  const extra = hasHalf ? 1 : 0

  return filled.repeat(full) + (hasHalf ? half : '') + empty.repeat(safeWidth - full - extra)
}

/** Renders a context meter with the raw usage percentage beside it. */
export const contextBarWithPct = ({ showPct = true, used, window, ...options }: ContextBarWithPctOptions): string => {
  const bar = contextBar({ used, window, ...options })

  if (!showPct) {
    return bar
  }

  const pct = window > 0 ? (used / window) * 100 : 0

  return `${bar} ${pct.toFixed(1).padStart(5)}%`
}
